### DISCUSSION TITLE
Surprising behavior with dynamic taxes sync webhook

### DISCUSSION THREAD

**Discussion by rrrnld** [saleor/saleor#10732] [Q&A]
Hi,

I'm trying to migrate to calculating taxes in an app because the changes in 3.6 broke the plugin I wrote for that. I wrote some tests, but it seems that my tax calculations are ignored and `checkout.totalPrice.net` is never updated. I think the body that I return from my webhook handler might have a wrong shape, which causes saleor to ignore it.

What exactly is the request body of the webhook response (the part that I implement in an app) supposed to look like, did I make a mistake there? I looked at https://docs.saleor.io/docs/3.x/developer/extending/apps/synchronous-webhooks/tax-webhooks#response-format as a guidance. `example-request.json` is a file that contains a webhook request to my handler that I intercepted.

```
# curlie POST :8081/handler -d @./example-request.json
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Mon, 26 Sep 2022 11:49:16 GMT
Content-Length: 177

{
    "shipping_tax_rate": "0",
    "shipping_price_gross_amount": "0",
    "shipping_price_net_amount": "0",
    "lines": [
        {
            "total_net_amount": "36.44859813084112",
            "total_gross_amount": "39",
            "tax_rate": "7"
        }
    ]
}
```

I tried returning an array (because the request itself also comes as an array of a single checkout), but the `total_net_amount` always seems to get ignored.

One thing I'd like to add: The `*_amount` fields are described as floats in the docs. I tried returning floats to no avail, and i think the amounts are expected to be returned as strings for precision, correct? In that case the documentation is incorrect.

Any help would be very appreciated!

**EDIT** It seems the response body is supposed to look exactly like this. However, it seems like the value is rounded before it's used for further calculation (see https://github.com/saleor/saleor/discussions/10732#discussioncomment-3734747), is that correct?

**Comment by rrrnld** [comment]
Ooops, my mistake. Using the above webhook response the taxes returned from the app are indeed taken into consideration while calculating taxes. I have however noticed other surprising behavior and I'm not sure what causes it.

I create a checkout using items with different tax rates. The `createCheckout` response looks like this:

<details>
<summary>GraphQL mutation response</summary>

``` json
{
  "checkoutCreate": {
    "checkout": {
      "id": "Q2hlY2tvdXQ6ZTcyZTk3YTktMDkxZi00NjQ0LTlmNjAtMTVlN2I0N2Y1ZTkx",
      "token": "e72e97a9-091f-4644-9f60-15e7b47f5e91",
      "totalPrice": {
        "gross": {
          "amount": 195,
          "currency": "EUR"
        },
        "net": {
          "amount": 174.9,
          "currency": "EUR"
        }
      },
      "lines": [
        {
          "variant": {
            "id": "UHJvZHVjdFZhcmlhbnQ6MTg3Ng==",
            "product": {
              "name": "Some name",
              "productType": {
                "name": "Type 1"
              }
            },
            "metadata": []
          },
          "unitPrice": {
            "gross": {
              "amount": 39,
              "currency": "EUR"
            },
            "net": {
              "amount": 32.78,
              "currency": "EUR"
            }
          },
          "totalPrice": {
            "gross": {
              "amount": 78,
              "currency": "EUR"
            },
            "net": {
              "amount": 65.55,
              "currency": "EUR"
            }
          }
        },
        {
          "variant": {
            "id": "UHJvZHVjdFZhcmlhbnQ6MTg3OA==",
            "product": {
              "name": "Some other name",
              "productType": {
                "name": "Type 2"
              }
            },
            "metadata": []
          },
          "unitPrice": {
            "gross": {
              "amount": 39,
              "currency": "EUR"
            },
            "net": {
              "amount": 36.45,
              "currency": "EUR"
            }
          },
          "totalPrice": {
            "gross": {
              "amount": 117,
              "currency": "EUR"
            },
            "net": {
              "amount": 109.35,
              "currency": "EUR"
            }
          }
        }
      ],
      "isShippingRequired": true,
      "shippingMethods": [
        {
          "id": "U2hpcHBpbmdNZXRob2Q6Mw==",
          "name": "Some shipping method",
          "active": true,
          "message": ""
        }
      ],
      "availableCollectionPoints": [],
      "availablePaymentGateways": [
        {
          "id": "saleor.payments.stripe",
          "name": "Stripe",
          "config": [
            {
              "field": "api_key",
              "value": "api_key_value"
            }
          ]
        }
      ],
      "metadata": []
    },
    "errors": []
  }
}

```
</details>

However, my tax calculation for this checkout looks like this:

``` json
{
    "shipping_tax_rate": "0.07",
    "shipping_price_gross_amount": "0",
    "shipping_price_net_amount": "0",
    "lines": [
        {
            "total_net_amount": "65.546218487394958",
            "total_gross_amount": "78",
            "tax_rate": "19"
        },
        {
            "total_net_amount": "109.3457943925233645",
            "total_gross_amount": "117",
            "tax_rate": "7"
        }
    ]
}
```

As you can see the net amounts are all rounded, and as a result the total net amount appears to be slightly off. It should be 174.892013 (or 174.89), but it's 174.9. Is that a bug?

**Reply by patrys** [reply]
All amounts are quantized (using bank rounding) to the number of decimal places correct for the currency used. For USD that's two places.

1. `65.546218487394958` becomes `65.55`
2. `109.3457943925233645` becomes `109.35`
3. The sum of the two is `174.90`

**Comment by TomSteer** [comment]
We also seem to be experiencing issues with the tax webhook. It hits our webhook endpoint, and we are returning JSON in the structure @heyarne mentions in his example (tried as both string and decimals). However, it seems to be ignored, and the untaxed product price is used for the order line total (gross & net).

**Reply by rrrnld** [reply]
The response I posted in the first post actually works and the tax rates I calculate are used in the response (albeit slightly wrong because apparently rounded values are used?). Since the webhook request from saleor looks like `[{"type": "Checkout", ...}]`, I simply use `checkouts[0]` to calculate the taxes.

**Reply by TomSteer** [reply]
👍 Looks like it was an issue due to the webhook proxy we were using. Now seems to be working as expected.
